ntoskrnl/ke/i386/mp.c

```diff
@@ -0,0 +1,127 @@
+/*
+ * PROJECT:     ReactOS Kernel
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Architecture specific source file to hold multiprocessor functions
+ * COPYRIGHT:   Copyright 2021 Justin Miller <justinmiller100@gmail.com>
+ */
+
+/* INCLUDES *****************************************************************/
+
+#include <ntoskrnl.h>
+#define NDEBUG
+#include <debug.h>
+
+/* GLOBALS *******************************************************************/
+
+#define PHYSICAL_ADDRESS  LARGE_INTEGER
+PHYSICAL_ADDRESS HighestPhysicalAddress;
+ULONG ProcessorCount = 0;
+
+struct APInfo
+{
+    KPCR pcr;
+    KTSS tss;
+    ETHREAD thread;
+    KDESCRIPTOR GdtDesc;
+    KDESCRIPTOR IdtDesc;
+} APInfo;
+
+/* FUNCTIONS *****************************************************************/
+VOID
+NTAPI
+KeStartAllProcessors()
+{
+     /* Prep some globals for the AP */
+    PVOID KernelStack;
+    PVOID DPCStack;
+    KPROCESSOR_STATE ProcessorState;
+    KDESCRIPTOR GdtDesc, IdtDesc;
+
+    HighestPhysicalAddress.QuadPart = UINT64_MAX;
+    __sgdt(&GdtDesc.Limit);
+    __sidt(&IdtDesc.Limit);
+
+    /* Attempt to allocate memory for the new setup */
+    SIZE_T APInfoSize = sizeof(struct APInfo) + 2;
+    PVOID PAPInfo = MmAllocateContiguousMemory(APInfoSize, HighestPhysicalAddress);
+    ASSERT(PAPInfo);
+
+    do
+    {
+        ProcessorCount++;
+
+        RtlCopyMemory((PVOID)((ULONG_PTR)PAPInfo + sizeof(KPCR) + sizeof(KTSS)), (PVOID)GdtDesc.Base, GdtDesc.Limit + 1);
+        RtlCopyMemory((PVOID)((ULONG_PTR)PAPInfo + sizeof(KPCR) + sizeof(KTSS) + GdtDesc.Limit + 1), (PVOID)IdtDesc.Base, IdtDesc.Limit + 1);
+        
+        ProcessorState.SpecialRegisters.Gdtr.Limit = GdtDesc.Limit;
+        ProcessorState.SpecialRegisters.Gdtr.Base = (ULONG)((ULONG_PTR)PAPInfo + sizeof(KPCR) + sizeof(KTSS));
+        ProcessorState.SpecialRegisters.Idtr.Limit = IdtDesc.Limit;
+        ProcessorState.SpecialRegisters.Idtr.Base = (ULONG)((ULONG_PTR)PAPInfo + sizeof(KPCR) + sizeof(KTSS) + GdtDesc.Limit + 1);
+
+        /* Prep Cr Regsters */
+        ProcessorState.SpecialRegisters.Cr0 = __readcr0();
+        ProcessorState.SpecialRegisters.Cr2 = __readcr2();
+        ProcessorState.SpecialRegisters.Cr3 = __readcr3();
+        ProcessorState.SpecialRegisters.Cr4 = __readcr4();
+
+          /* Prepare Segment Registers */
+        ProcessorState.ContextFrame.SegCs = KGDT_R0_CODE;
+        ProcessorState.ContextFrame.SegSs = KGDT_R0_DATA;
+        ProcessorState.ContextFrame.SegDs = KGDT_R0_DATA;
+        ProcessorState.ContextFrame.SegEs = KGDT_R0_DATA;
+
+        /* Clear GS */
+        ProcessorState.ContextFrame.SegGs = 0;
+        /* Set FS to PCR */
+        ProcessorState.ContextFrame.SegFs = KGDT_R0_PCR;
+
+        /* Load TSR */
+        ProcessorState.SpecialRegisters.Tr = KGDT_TSS;
+
+        ProcessorState.ContextFrame.Eip = (ULONG_PTR)KiSystemStartup;
```

Comment:
I don't think any of this is needed here, since the values are hardcoded and can be loaded in APSpinup as well.

Reply:
The NT kernel is responsible for telling the HAL what state to initialize the application processors. The HAL just starts the processors and place them in whatever state the kernel told it to use. In particular the kernel knows what protected mode entrypoint the application processors need to execute ultimately.

Reply:
Ah, so this is Windows compatibility thing. Fair enough.